Add dataset_types to package info

ruflin opened the pull request:
To be updated

elasticmachine commented:
## :broken_heart: Build Failed


<!-- BUILD BADGES-->
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fpackage-registry%2FPR-511/detail/PR-511/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fpackage-registry%2FPR-511/detail/PR-511/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fpackage-registry%2FPR-511/detail/PR-511/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fpackage-registry%2FPR-511/detail/PR-511/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://package-registry_511.docs-preview.app.elstc.co/diff)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats

* Build Cause: `[Pull request #511 opened]`

* Start Time: 2020-06-15T13:48:11.759+0000

* Duration: 3 min 7 sec





</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY-->


### Steps errors

  <details><summary>Expand to view the steps failures</summary>
  <p>
  
  * **Name**: `Checks formatting / linting`
    * Description: mage -debug check  
    
    * Duration: 0 min 9 sec
    
    * Start Time: 2020-06-15T13:51:06.701+0000
    
    * [log](https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/package-registry/pipelines/PR-511/runs/1/steps/91/log)
    
  
  </p>
  </details>



<!-- LOG OUTPUT-->
### Log output

<details><summary>Expand to view the last 100 lines of log output</summary>
<p>

```
[2020-06-15T13:51:04.422Z] .
[2020-06-15T13:51:04.523Z] beats-ci-immutable-ubuntu-1604-1592228907484540473 does not seem to be running inside a container
[2020-06-15T13:51:04.555Z] $ docker run -t -d -u 1156:1157 -w /var/lib/jenkins/workspace/Beats_package-registry_PR-511/src/github.com/elastic/package-registry -v /var/lib/jenkins/workspace/Beats_package-registry_PR-511/src/github.com/elastic/package-registry:/var/lib/jenkins/workspace/Beats_package-registry_PR-511/src/github.com/elastic/package-registry:rw,z -v /var/lib/jenkins/workspace/Beats_package-registry_PR-511/src/github.com/elastic/package-registry@tmp:/var/lib/jenkins/workspace/Beats_package-registry_PR-511/src/github.com/elastic/package-registry@tmp:rw,z -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** -e ******** go-agent cat
[2020-06-15T13:51:05.321Z] $ docker top 3ac4fb81ca5bfcfb8e941707175c72b9fee6c49b45911ca1ac1bd3ac7875e4cc -eo pid,comm
[2020-06-15T13:51:06.070Z] + go version
[2020-06-15T13:51:06.070Z] go version go1.14.2 linux/amd64
[2020-06-15T13:51:06.687Z] + mage -version
[2020-06-15T13:51:06.687Z] Mage Build Tool <not set>
[2020-06-15T13:51:06.687Z] Build Date: <not set>
[2020-06-15T13:51:06.687Z] Commit: <not set>
[2020-06-15T13:51:06.687Z] built with: go1.14.2
[2020-06-15T13:51:07.306Z] + mage -debug check
[2020-06-15T13:51:07.306Z] DEBUG: 13:51:07.053836 getting all non-mage files in .
[2020-06-15T13:51:07.568Z] DEBUG: 13:51:07.349025 found non-mage files artifacts.go||categories.go||favicon.go||handler.go||index.go||main.go||mime_types.go||package_index.go||search.go||types.go
[2020-06-15T13:51:07.568Z] DEBUG: 13:51:07.349067 marked file as non-mage: "artifacts.go"
[2020-06-15T13:51:07.568Z] DEBUG: 13:51:07.349074 marked file as non-mage: "categories.go"
[2020-06-15T13:51:07.568Z] DEBUG: 13:51:07.349078 marked file as non-mage: "favicon.go"
[2020-06-15T13:51:07.568Z] DEBUG: 13:51:07.349082 marked file as non-mage: "handler.go"
[2020-06-15T13:51:07.568Z] DEBUG: 13:51:07.349087 marked file as non-mage: "index.go"
[2020-06-15T13:51:07.568Z] DEBUG: 13:51:07.349091 marked file as non-mage: "main.go"
[2020-06-15T13:51:07.568Z] DEBUG: 13:51:07.349096 marked file as non-mage: "mime_types.go"
[2020-06-15T13:51:07.568Z] DEBUG: 13:51:07.349100 marked file as non-mage: "package_index.go"
[2020-06-15T13:51:07.568Z] DEBUG: 13:51:07.349104 marked file as non-mage: "search.go"
[2020-06-15T13:51:07.568Z] DEBUG: 13:51:07.349117 marked file as non-mage: "types.go"
[2020-06-15T13:51:07.568Z] DEBUG: 13:51:07.349120 getting all files plus mage files
[2020-06-15T13:51:08.141Z] DEBUG: 13:51:08.066327 time to scan for Magefiles: 1.012554531s
[2020-06-15T13:51:08.141Z] DEBUG: 13:51:08.066352 found magefiles: magefile.go
[2020-06-15T13:51:08.141Z] DEBUG: 13:51:08.072486 output exe is  /var/lib/jenkins/workspace/Beats_package-registry_PR-511/src/github.com/elastic/package-registry/.magefile/4abaaedc8622d69edcfebc5112be06d4f331e5bf
[2020-06-15T13:51:08.141Z] DEBUG: 13:51:08.081183 go build cache exists, will ignore any compiled binary
[2020-06-15T13:51:08.141Z] DEBUG: 13:51:08.081204 parsing files
[2020-06-15T13:51:08.141Z] DEBUG: 13:51:08.082049 found target AddLicenseHeaders
[2020-06-15T13:51:08.141Z] DEBUG: 13:51:08.082077 found target Build
[2020-06-15T13:51:08.141Z] DEBUG: 13:51:08.082081 found target Check
[2020-06-15T13:51:08.141Z] DEBUG: 13:51:08.082085 found target Clean
[2020-06-15T13:51:08.141Z] DEBUG: 13:51:08.082105 skipping function with invalid signature func FindFilesRecursive(match &{4781 0xc0002a8480 0xc0002a84b0})(&{4824 <nil> string}, error)
[2020-06-15T13:51:08.141Z] DEBUG: 13:51:08.082108 found target Format
[2020-06-15T13:51:08.141Z] DEBUG: 13:51:08.082121 found target GoImports
[2020-06-15T13:51:08.141Z] DEBUG: 13:51:08.082132 found target PrepareTest
[2020-06-15T13:51:08.141Z] DEBUG: 13:51:08.082136 found target Test
[2020-06-15T13:51:08.141Z] DEBUG: 13:51:08.082139 found target TestIntegration
[2020-06-15T13:51:08.141Z] DEBUG: 13:51:08.082143 found target Vendor
[2020-06-15T13:51:08.141Z] DEBUG: 13:51:08.082156 time parse Magefiles: 946.033µs
[2020-06-15T13:51:08.141Z] DEBUG: 13:51:08.082171 Creating mainfile at mage_output_file.go
[2020-06-15T13:51:08.141Z] DEBUG: 13:51:08.082221 writing new file at mage_output_file.go
[2020-06-15T13:51:08.141Z] DEBUG: 13:51:08.083580 compiling to /var/lib/jenkins/workspace/Beats_package-registry_PR-511/src/github.com/elastic/package-registry/.magefile/4abaaedc8622d69edcfebc5112be06d4f331e5bf
[2020-06-15T13:51:08.141Z] DEBUG: 13:51:08.083589 compiling using gocmd: go
[2020-06-15T13:51:08.403Z] DEBUG: 13:51:08.326815 running go build -o /var/lib/jenkins/workspace/Beats_package-registry_PR-511/src/github.com/elastic/package-registry/.magefile/4abaaedc8622d69edcfebc5112be06d4f331e5bf magefile.go mage_output_file.go
[2020-06-15T13:51:09.349Z] DEBUG: 13:51:09.030630 time to compile Magefile: 703.755251ms
[2020-06-15T13:51:09.349Z] DEBUG: 13:51:09.030696 running binary /var/lib/jenkins/workspace/Beats_package-registry_PR-511/src/github.com/elastic/package-registry/.magefile/4abaaedc8622d69edcfebc5112be06d4f331e5bf
[2020-06-15T13:51:09.349Z] DEBUG: 13:51:09.030712 running magefile with mage vars:
[2020-06-15T13:51:09.349Z] MAGEFILE_DEBUG=1
[2020-06-15T13:51:09.349Z] MAGEFILE_GOCMD=go
[2020-06-15T13:51:09.349Z] >> fmt - go-licenser: Adding missing headers
[2020-06-15T13:51:09.349Z] >> fmt - goimports: Formatting Go code
[2020-06-15T13:51:09.349Z] Cloning into 'build/package-storage'...
[2020-06-15T13:51:10.735Z] Already on 'master'
[2020-06-15T13:51:12.660Z] 2020/06/15 13:51:12 >> Copy package: testdata/package/datasources/1.0.0/
[2020-06-15T13:51:12.660Z] 2020/06/15 13:51:12 exec: rsync -a testdata/package/datasources/1.0.0/ testdata/public/package/datasources/1.0.0
[2020-06-15T13:51:12.660Z] 2020/06/15 13:51:12 >> Copy package: testdata/package/default-pipeline/0.0.2/
[2020-06-15T13:51:12.660Z] 2020/06/15 13:51:12 exec: rsync -a testdata/package/default-pipeline/0.0.2/ testdata/public/package/default-pipeline/0.0.2
[2020-06-15T13:51:12.660Z] 2020/06/15 13:51:12 >> Copy package: testdata/package/example/0.0.2/
[2020-06-15T13:51:12.660Z] 2020/06/15 13:51:12 exec: rsync -a testdata/package/example/0.0.2/ testdata/public/package/example/0.0.2
[2020-06-15T13:51:12.660Z] 2020/06/15 13:51:12 >> Copy package: testdata/package/example/1.0.0/
[2020-06-15T13:51:12.660Z] 2020/06/15 13:51:12 exec: rsync -a testdata/package/example/1.0.0/ testdata/public/package/example/1.0.0
[2020-06-15T13:51:12.660Z] 2020/06/15 13:51:12 >> Copy package: testdata/package/experimental/0.0.1/
[2020-06-15T13:51:12.660Z] 2020/06/15 13:51:12 exec: rsync -a testdata/package/experimental/0.0.1/ testdata/public/package/experimental/0.0.1
[2020-06-15T13:51:12.660Z] 2020/06/15 13:51:12 >> Copy package: testdata/package/foo/1.0.0/
[2020-06-15T13:51:12.660Z] 2020/06/15 13:51:12 exec: rsync -a testdata/package/foo/1.0.0/ testdata/public/package/foo/1.0.0
[2020-06-15T13:51:12.660Z] 2020/06/15 13:51:12 >> Copy package: testdata/package/internal/1.2.0/
[2020-06-15T13:51:12.660Z] 2020/06/15 13:51:12 exec: rsync -a testdata/package/internal/1.2.0/ testdata/public/package/internal/1.2.0
[2020-06-15T13:51:12.660Z] 2020/06/15 13:51:12 >> Copy package: testdata/package/multiple-false/0.0.1/
[2020-06-15T13:51:12.660Z] 2020/06/15 13:51:12 exec: rsync -a testdata/package/multiple-false/0.0.1/ testdata/public/package/multiple-false/0.0.1
[2020-06-15T13:51:12.660Z] 2020/06/15 13:51:12 >> Copy package: testdata/package/no_stream_configs/1.0.0/
[2020-06-15T13:51:12.660Z] 2020/06/15 13:51:12 exec: rsync -a testdata/package/no_stream_configs/1.0.0/ testdata/public/package/no_stream_configs/1.0.0
[2020-06-15T13:51:14.576Z] >> mod - updating vendor directory
[2020-06-15T13:51:14.576Z] go: downloading github.com/stretchr/testify v1.4.0
[2020-06-15T13:51:14.576Z] go: downloading github.com/davecgh/go-spew v1.1.0
[2020-06-15T13:51:14.576Z] go: downloading gopkg.in/check.v1 v0.0.0-20161208181325-20d25e280405
[2020-06-15T13:51:14.576Z] go: downloading github.com/pmezard/go-difflib v1.0.0
[2020-06-15T13:51:14.836Z] all modules verified
[2020-06-15T13:51:15.097Z] go: downloading github.com/jstemmer/go-junit-report v0.9.1
[2020-06-15T13:51:15.097Z] go: github.com/jstemmer/go-junit-report upgrade => v0.9.1
[2020-06-15T13:51:15.097Z] github.com/jstemmer/go-junit-report/parser
[2020-06-15T13:51:15.358Z] github.com/jstemmer/go-junit-report/formatter
[2020-06-15T13:51:15.358Z] github.com/jstemmer/go-junit-report
[2020-06-15T13:51:15.618Z] types.go: needs update
[2020-06-15T13:51:15.618Z] Error: running "git update-index --refresh" failed with exit code 1
[2020-06-15T13:51:15.633Z] $ docker stop --time=1 3ac4fb81ca5bfcfb8e941707175c72b9fee6c49b45911ca1ac1bd3ac7875e4cc
[2020-06-15T13:51:17.366Z] $ docker rm -f 3ac4fb81ca5bfcfb8e941707175c72b9fee6c49b45911ca1ac1bd3ac7875e4cc
[2020-06-15T13:51:17.610Z] Stage "Build" skipped due to earlier failure(s)
[2020-06-15T13:51:17.677Z] Stage "Test" skipped due to earlier failure(s)
[2020-06-15T13:51:17.748Z] Stage "TestIntegration" skipped due to earlier failure(s)
[2020-06-15T13:51:17.821Z] Stage "Publish Docker image" skipped due to earlier failure(s)
[2020-06-15T13:51:18.042Z] Running on Jenkins in /var/lib/jenkins/workspace/Beats_package-registry_PR-511
[2020-06-15T13:51:18.159Z] [INFO] getVaultSecret: Getting secrets
[2020-06-15T13:51:18.218Z] Masking supported pattern matches of $VAULT_ADDR or $VAULT_ROLE_ID or $VAULT_SECRET_ID
[2020-06-15T13:51:18.988Z] + chmod 755 generate-build-data.sh
[2020-06-15T13:51:18.988Z] + ./generate-build-data.sh https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/package-registry/PR-511/ https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/package-registry/PR-511/runs/1 FAILURE 186960
[2020-06-15T13:51:18.988Z] INFO: curl https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/package-registry/PR-511/runs/1/steps/?limit=10000 -o steps-info.json
[2020-06-15T13:51:19.239Z] INFO: curl https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/package-registry/PR-511/runs/1/tests/?status=FAILED -o tests-errors.json

```

</p>
</details>



<!--COMMENT_GENERATED_WITH_ID_comment.id-->

ruflin commented:
As this does not remove / change the /categories endpoint yet, this can be merged and the change in Kibana can follow. Issue filed here: https://github.com/elastic/kibana/issues/69252

mtojek commented on types.go:
```diff
@@ -0,0 +1,119 @@
+
+		data, err := getTypesOutput(types)
+		if err != nil {
+			notFoundError(w, err)
```
I think you can get here a marshaling error. It would be great if the app logs this one.

mtojek commented on util/package.go:
```diff
@@ -50,6 +50,8 @@ type Package struct {
 	Datasources   []Datasource `config:"datasources,omitempty" json:"datasources,omitempty" yaml:"datasources,omitempty"`
 	Download      string       `json:"download" yaml:"download,omitempty"`
 	Path          string       `json:"path" yaml:"path,omitempty"`
+	// This is in purpose `dataset.types` with an "s" as it is a summary of the existing dataset types
```
nit: on purpose:

ruflin replied:
fixed

mtojek commented on util/package.go:
```diff
@@ -50,6 +50,8 @@ type Package struct {
 	Download      string       `json:"download" yaml:"download,omitempty"`
 	Path          string       `json:"path" yaml:"path,omitempty"`
+	// This is in purpose `dataset.types` with an "s" as it is a summary of the existing dataset types
+	DatasetTypes []string `json:"dataset.types" yaml:"dataset.types,omitempty"`
```
should both ones have the `omitempty` annotation?

ruflin replied:
yes, and I realised that we don't need the yaml part as it never shows up in yaml.

mtojek commented on util/package.go:
```diff
@@ -452,3 +473,20 @@ func (p *Package) GetDownloadPath() string {
 	return path.Join("/package", p.Name, p.Version)
 }
+
+// getDatasetTypes return all the dataset types which exist
```
nit: returns

ruflin replied:
fixed

mtojek commented on util/package.go:
```diff
@@ -452,3 +473,20 @@ func (p *Package) GetDownloadPath() string {
+	// Get unique list of
+	typesCounter := map[string]int{}
+	for _, d := range p.DataSets {
+		typesCounter[d.Type] = typesCounter[d.Type] + 1
```
I wonder if it doesn't break on the first execution:
`typesCounter[d.Type] = typesCounter[d.Type] + 1`

`typesCounter` is empty, so `typesCounter[d.Type]` would fail.

ruflin replied:
Glad you stumbled over this. I remember thinking exactly the same but then forgetting up it because it works. Need to check in more details.

ruflin replied:
fixed

ruflin commented on docs/api/package.json:
```diff
@@ -116,5 +116,8 @@
   "download": "/epr/example/example-1.0.0.tar.gz",
-  "path": "/package/example/1.0.0"
+  "path": "/package/example/1.0.0",
+  "dataset.types": [
```
This is by design to be as close to `dataset.type`. I probably best make it an object.

ruflin replied:
Changed to an object

ruflin commented on main.go:
```diff
@@ -148,6 +148,7 @@ func getRouter(config Config, packagesBasePath string) (*mux.Router, error) {
 	router.HandleFunc("/index.json", indexHandlerFunc)
 	router.HandleFunc("/search", searchHandler(packagesBasePath, config.CacheTimeSearch))
 	router.HandleFunc("/categories", categoriesHandler(packagesBasePath, config.CacheTimeCategories))
+	router.HandleFunc("/types", typesHandler(packagesBasePath, config.CacheTimeCategories))
```
You mean it will conflict because of packagesBasePath? SGTM, will do the rebase fun.

mtojek replied:
No, my point is just to inform you about it :D 

I will take care of it. I don't expect it to be huge refactoring.

ruflin commented on search.go:
```diff
@@ -48,6 +49,12 @@ func searchHandler(packagesBasePath string, cacheTime time.Duration) func(w http
 				}
 			}
 
+			if v := query.Get("dataset.type"); v != "" {
```
This is by design. I want to stay as close as possible to the actual fields used in the events. So the query is really `dataset.type=logs` like it is inside Elasticsearch.

ruflin commented on types.go:
```diff
@@ -0,0 +1,119 @@
+// typesHandler
+func typesHandler(packagesBasePath string, cacheTime time.Duration) func(w http.ResponseWriter, r *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		packages, err := util.GetPackages(packagesBasePath)
```
What you mean by separate goroutine here? Are you worried about the performance?

ruflin commented on docs/api/search-datasetType-logs.json:
```diff
@@ -0,0 +1,47 @@
+[
+  {
+    "description": "Package with data sources",
```
fixed

ruflin commented on docs/api/search-datasetType-logs.json:
```diff
@@ -0,0 +1,47 @@
+    "version": "1.0.0"
+  },
+  {
+    "description": "Tests if no pipeline is set, it defaults to the default one",
```
updated

ruflin commented on search.go:
```diff
@@ -48,6 +49,12 @@ func searchHandler(packagesBasePath string, cacheTime time.Duration) func(w http
 			}
 
+			if v := query.Get("dataset.type"); v != "" {
+				if v != "" {
```
fixed

ruflin commented on search.go:
```diff
@@ -108,6 +115,13 @@ func searchHandler(packagesBasePath string, cacheTime time.Duration) func(w http
 				continue
 			}
 
+			// Filter by dataset.type first as this could heavily reduce the number of packages
```
fixed

ruflin commented:
Waiting for https://github.com/elastic/package-registry/pull/529 to get in to continue work on this PR.

I converted above `dataset.types` to `dataset: { types` but I'm torn if this is the right call. The reason I converted it to an object is to align with `dataset.*` object but at the same time inside this object there is not `types` (plural) and it kind of implies a dataset could have more than one type (it can't). So I'm getting back to the proposal from @mtojek to use `dataset_types`. Will adjust it.

ruflin commented:
I'm closing this for now as it turned out we will not need it yet for the next release.
